program/rust/src/log.rs

```diff
@@ -81,24 +100,51 @@ pub fn pre_log(accounts : &[AccountInfo], instruction_data: &[u8]) -> ProgramRes
 
 pub fn post_log(c_ret_val: u64, accounts: &[AccountInfo]) -> ProgramResult {
     if c_ret_val == SUCCESSFULLY_UPDATED_AGGREGATE {
-        let start: usize = PRICE_T_AGGREGATE_OFFSET
+        let aggregate_price_start: usize = PRICE_T_AGGREGATE_OFFSET
             .try_into()
             .map_err(|_| OracleError::Generic)?;
         // We trust that the C oracle has properly checked this account
         let aggregate_price_info: pc_price_info = pc_price_info::try_from_slice(
             &accounts
                 .get(1)
                 .ok_or(OracleError::Generic)?
-                .try_borrow_data()?[start..(start + size_of::<pc_price_info>())],
+                .try_borrow_data()?[aggregate_price_start..(aggregate_price_start + size_of::<pc_price_info>())],
         )?;
+
+        let ema_start: usize = PRICE_T_EMA_OFFSET
+            .try_into()
+            .map_err(|_| OracleError::Generic)?;
+
+        let ema_info: pc_ema = pc_ema::try_from_slice(
+            &accounts
+                .get(1)
+                .ok_or(OracleError::Generic)?
+                .try_borrow_data()?[ema_start..(ema_start + size_of::<pc_ema>())],
+        )?;
+
+        let expo_start : usize = PRICE_T_EXPO_OFFSET.try_into()
+        .map_err(|_| OracleError::Generic)?;
+
+        let expo: i32 = i32::try_from_slice(
+            &accounts
+                .get(1)
+                .ok_or(OracleError::Generic)?
```

Comment:
Instead of generic errors, could we create actual errors here? You can use the `anyhow:anyhow` macro to create "ad-hoc" errors from strings: https://docs.rs/anyhow/latest/anyhow/macro.anyhow.html

Reply:
Made actual errors in the enum